### DISCUSSION TITLE
Electron NodeJS files

### DISCUSSION THREAD

**Discussion by PixelForm** [guasam/electron-react-app#32] [Q&A] [answered]
First off, this boilerplate is perfect!

I just can't seem to find where I put my scripts for sharing data between NodeJS and my React UI. I rather keep the NodeIntegration on false and use something like the electron docs say, a preload.js kind of method. I wonder if the bridge between Node and React has maybe already been made by the boilerplate but I cannot find it if that's the case. I simply need to "require()" some NodeJs modules and securely expose the output from those script to my React UI.

I don't see anything like this in the Readme either so please help.

**Comment by codesbiome** [comment, marked as answer]
Hello @PixelForm

> First off, this boilerplate is perfect!

Your welcome!

> I just can't seem to find where I put my scripts for sharing data between NodeJS and my React UI. I rather keep the NodeIntegration on false and use something like the electron docs say, a preload.js kind of method. I wonder if the bridge between Node and React has maybe already been made by the boilerplate but I cannot find it if that's the case. I simply need to "require()" some NodeJs modules and securely expose the output from those script to my React UI.

To share data between `main` and `renderer` process you can use electron's  [contextBridge](https://www.electronjs.org/docs/v14-x-y/api/context-bridge) api.

We used the same api in this project for the custom window frame/titlebar events handling.
You can look some implementation of context bridge in  `misc/window` directory, for example: 
```js
// misc/window/windowPreload.ts

import { contextBridge } from 'electron';
import titlebarContext from './titlebarContext';

contextBridge.exposeInMainWorld('electron_window', {
  titlebar: titlebarContext,
});
```
Exposed api by context-bridge will be available in `renderer` process (browser), so the api mentioned above is also available to use in Developer Tools (console): 

![image](https://user-images.githubusercontent.com/46080054/180825881-5515d9a7-ce53-4f08-b51c-179bba296f93.png)


All window frame/titlebar specific events are defined using IPC api, for example : 

```js
misc/window/titlebarIPC.ts  // Exposing events in renderer process (browser) for window titlebar buttons, menus etc.
misc/window/titlebarIPC.ts  // Defining callbacks for main process to handle all requested window events of renderer process.
```

**Reply by PixelForm** [reply]
Wonderful! I'll look into it.
